pallets/collator-rewards/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		sp_runtime::traits::AccountIdConversion,
+		traits::{Currency, ExistenceRequirement, WithdrawReasons},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+
+	pub type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	pub type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+		<T as frame_system::Config>::AccountId,
+	>>::NegativeImbalance;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The currency implementation of the runtime
+		type Currency: Currency<Self::AccountId>;
+		/// The origin permitted to set the reward amount
+		type ForceOrigin: EnsureOrigin<Self::Origin>;
+		/// PalletId for this pallet
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+	}
+
+	/// The RewardAmount to be paid out every block to author
+	#[pallet::storage]
+	#[pallet::getter(fn reward_amount)]
+	pub type RewardAmount<T> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Reward Amount set to value
+		RewardAmountSet(BalanceOf<T>),
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set the reward amount to be paid out every block to collators
+		/// Can only be called by the ForceOrigin
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn force_set_reward_amount(
+			origin: OriginFor<T>,
+			reward_amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			T::ForceOrigin::ensure_origin(origin)?;
+			// Update storage.
+			<RewardAmount<T>>::put(reward_amount);
+			// Emit an event.
+			Self::deposit_event(Event::RewardAmountSet(reward_amount));
+			// Return a successful DispatchResultWithPostInfo
+			Ok(().into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Withdraw the reward to pay to the collator account.
+		/// This function returns a NegativeImbalance as the result of withdrawal/issuance. The
+		/// pallet account is checked for balance to process this withdrawal. If the withdrawal from
+		/// pallet account fails, new issuance is created.
+		pub fn withdraw_reward() -> Result<NegativeImbalanceOf<T>, DispatchError> {
+			// check if the pallet account has the balance to withdraw the amount
+			// else we issue the newly created amount
+			match T::Currency::withdraw(
```

Comment:
Is there any way this fails even with the balance existing? I.e returns an Err when there is still non-zero funds in the pot?

Reply:
I looked into this, the only way this would fail at non-zero balance is if there is some sort of lock/reserve on the balance, don't think that would happen here (or we fail to allot balance correctly in genesis). For the abuse part, not sure there's an opportunity here, we are using the same logic as polkadot to pay the block reward fee, just adding a new amount to it, afaik there is no way to do this more than once per block. But outside of it, I think there is an incentive to create empty blocks to be author to get the reward, although not sure how effective this would be under AURA